app/controllers/api/widget_sets_controller.rb

```diff
@@ -0,0 +1,134 @@
+module Api
+  class WidgetSetsController < BaseController
+    REQUIRED_FIELDS_TO_COPY = %w[name].freeze
+    ALLOWED_FIELDS = REQUIRED_FIELDS_TO_COPY + %w[group description guid read_only set_data].freeze
+    SET_DATA_COLS = %i[col1 col2 col3].freeze
+
+    def init_set_data(set_data)
+      set_data ||= {}
+      new_set_data ||= {}
+      new_set_data[:col1] = set_data['col1'] || []
+      new_set_data[:col2] = set_data['col2'] || []
+      new_set_data[:col3] = set_data['col3'] || []
+      new_set_data[:reset_upon_login] = !!set_data['reset_upon_login']
+      new_set_data[:locked] = !!set_data['locked']
+      new_set_data
+    end
```

Comment:
I'm confused with this, is the API defining a structure beyond the basic MiqWidgetSet ?   It almost feels we need a new subclass of MiqWidgetSet in the model for what we're building/managing here and move most of this logic in there.